src/DataTypes/FieldToDataType.cpp

```diff
@@ -36,6 +37,7 @@ DataTypePtr FieldToDataType<on_error>::operator() (const UInt64 & x) const
     if (x <= std::numeric_limits<UInt8>::max()) return std::make_shared<DataTypeUInt8>();
     if (x <= std::numeric_limits<UInt16>::max()) return std::make_shared<DataTypeUInt16>();
     if (x <= std::numeric_limits<UInt32>::max()) return std::make_shared<DataTypeUInt32>();
+    if (x <= std::numeric_limits<Int64>::max()) return std::make_shared<DataTypeUInt64WithOpposite>(std::make_shared<DataTypeInt64>());
```

Comment:
`DataTypeUInt64WithOpposite` type is returned, but this prevents passing through scenarios that require `typeid_cast<const DataTypeUInt64 *>(type)`, such as the `castBothTypes` method in `FunctionBinaryArithmetic`.